backend/src/main/java/co/inventorsoft/scripty/service/UserService.java

```diff
@@ -0,0 +1,91 @@
+import java.time.Clock;
+import java.time.Instant;
+import java.util.*;
+/**
+ *
+ * @author Symyniuk
+ *
+ */
+@Service
+@Transactional
+public class UserService {
+
+    private RoleRepository roleRepository;
+    private UserRepository userRepository;
+    private VerificationTokenRepository tokenRepository;
+    private PasswordEncoder passwordEncoder;
+    @Autowired
+    public UserService(RoleRepository roleRepository,
+                       UserRepository userRepository,
+                       VerificationTokenRepository tokenRepository,
+                       PasswordEncoder passwordEncoder){
+        this.roleRepository = roleRepository;
+        this.userRepository = userRepository;
+        this.tokenRepository = tokenRepository;
+        this.passwordEncoder = passwordEncoder;
+    }
+    public User findByEmail(final String email){
+        final User user = userRepository.findByEmail(email);
+        return user;
+    }
+
+    public User registerNewUserAccount(final UserDto userDto) {
+        if(findByEmail(userDto.getEmail()) != null){
+            throw new ApplicationException("There is an account with that email address: " +  userDto.getEmail(), HttpStatus.BAD_REQUEST);
+        }
+        final User user = new User();
+        user.setFirstName(userDto.getFirstName());
+        user.setLastName(userDto.getLastName());
+        user.setPassword(passwordEncoder.encode(userDto.getPassword()));
+        user.setEmail(userDto.getEmail());
+        user.setEnabled(false);
+        user.setRoles(Arrays.asList(roleRepository.findByName("User")));
+        return userRepository.save(user);
+    }
+
+    public void validateVerificationToken(final String token) {
+        Optional<VerificationToken> verificationTokenOptional = Optional.ofNullable(tokenRepository.findByToken(token));
+        if(!verificationTokenOptional.isPresent()){
+            throw new ApplicationException("Wrong link", HttpStatus.BAD_REQUEST);
+        }
+        final VerificationToken verificationToken = verificationTokenOptional.get();
+        final User user = verificationToken.getUser();
+        Instant instantExp = verificationToken.getExpiryDate();
+        final Instant instant = Clock.systemDefaultZone().instant();
+        if ((instantExp.isBefore(instant))) {
+            throw new ApplicationException("Time of user verification link has expired", HttpStatus.BAD_REQUEST);
+        }
+        user.setEnabled(true);
+        userRepository.save(user);
+        tokenRepository.delete(verificationToken);
+    }
+    public void createVerificationTokenForUser(final User user, final String token) {
+        final VerificationToken myToken = new VerificationToken(token, user);
+        tokenRepository.save(myToken);
+    }
+    public VerificationToken generateNewVerificationToken(final User user) {
+        Optional<VerificationToken> optionalToken = Optional.ofNullable(tokenRepository.findByUser(user));
```

Comment:
This can be replaced by Optional's composition methods. Example:
`
Optional.ofNullable(tokenRepository.findByUser(user))
.map(token -> token.updateToken(UUID.randomUUID().toString()))
.map(tokenRepository::save)
.orElseThrow(() -> new ApplicationException("Token not found", HttpStatus.OK))`

Note: this code will work if `VerificationToken.updateToken(String)` method returns `VerificationToken`. If this method returns `void` you need to wrap it in block and explicitly return token instance